### PR TITLE
Admin views

### PR DESCRIPTION
**Pull request by iamsitting**
#1165

### CODE REVIEW

**Comment by yipingbenjaminliu** [comment]
Cannot delete rolls attached to a leaveslip

**Comment by iamsitting** [comment]
@yipingbenjaminliu review

File: ap/attendance/views.py [hunk 1]
```diff
@@ -738,6 +738,16 @@ def form_valid(self, form):
...
 
+  def delete(self, request, *args, **kwargs):
+    self.object = self.get_object()
+
+    has_leaveslips = self.object.leaveslips.exists()
+
+    if not has_leaveslips:
+      return super(RollAdminDelete, self).delete(request, *args, **kwargs)
+    else:
+      raise Http404("This roll is attached to a leave slip. Please change to a present status instead of deleting it.")
```
Comment (yipingbenjaminliu) [inline ap/attendance/views.py]: change this to a form field error for the user to view rather than a 404 error?

**Comment by yipingbenjaminliu** [comment]
image
maybe just the red bar is enough

File: ap/attendance/forms.py [hunk 2]
```diff
@@ -22,6 +23,28 @@ def __init__(self, *args, **kwargs):
...
+
+    # Roll date and event needs to match schedule week
+    sched = data_event.schedules.filter(trainees__in=[data_trainee])
+    pr = sched.aggregate(Max('priority'))['priority__max']
+    if not sched.get(priority=pr).active_in_week(data_event.week_from_date(data_date)):
+      raise forms.ValidationError("Event is not active for date selected!!")
+
+    # trainees not on self attendance can only have one roll for any event on the same date
+    # trainees on self attendance can have two rolls for any event on the same date given that one is submitted by themselves and another one is not
+    # this will be enforced by unique_together on the database (PR #1243)
```
Comment (yipingbenjaminliu) [inline ap/attendance/forms.py]: actually, the unique_together won't enforce this

**Comment by iamsitting** [comment]
@yipingbenjaminliu what's next?

**Comment by yipingbenjaminliu** [comment]
@iamsitting, should be okay after this

**Pull request merged**